esp32/machine_i2c: Fix default I2C pins for C3 and S3 variants.

ricksorensen opened the pull request:
This addresses issue #17103.

The new default I2C init does not require setting SCL or SDA but the default pins for C3 (and S3?) conflict with the espressif GPIO usage.
For the C3, pins 18/19 are for USB/JTAG. If used for I2C() they will cause the REPL to hang on initialization of the I2C. For the S3 pin 19 is allocated for USB/JTAG also but the defaults do not seem to affect the REPL.

See PR #16956
Closes #17103

<!-- Thanks for submitting a Pull Request! We appreciate you spending the
     time to improve MicroPython. Please provide enough information so that
     others can review your Pull Request.

     Before submitting, please read:
     https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md
     https://github.com/micropython/micropython/wiki/ContributorGuidelines

     Please check any CI failures that appear after your Pull Request is opened.
-->

### Summary

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->


### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->


### Trade-offs and Alternatives

<!-- If the Pull Request has some negative impact (i.e. increased code size)
     then please explain why you think the trade-off improvement is worth it.
     If you can think of alternative ways to do this, please explain that here too.

     Delete this heading if not relevant (i.e. small fixes) -->

ricksorensen commented:
I have tested this on a UM TinyS3, SEEED XIAO S3, and SEEED XIAO C3.  I looked at the espressif documenation on GPIO for Esp32, C3,C6, S2, and S3 - and the only conflicts were in the C3 and S3,

Note also that most of the boards for esp32 define SCL and SDA in their mpconfigboard.h

dpgeorge commented:
Reviewing all C3 boards (there are only 2 of them) we have:
- ESP32_GENERIC_C3 does not define any I2C0 default pins, so will be affected by this PR
- LOLIN_C3_MINI does define:
```c
#define MICROPY_HW_I2C0_SCL                 (10)
#define MICROPY_HW_I2C0_SDA                 (8)
```

For S3 there are 11 boards:
- ARDUINO_NANO_ESP32 defines:
```c
#define MICROPY_HW_I2C0_SCL                 (12)
#define MICROPY_HW_I2C0_SDA                 (11)
```
- M5STACK_ATOMS3_LITE defines:
```c
#define MICROPY_HW_I2C0_SCL                 (39)
#define MICROPY_HW_I2C0_SDA                 (38)
```
- The remaining 9 boards (ESP32_GENERIC_S3 and all 8 UM S3 boards) define:
```c
#define MICROPY_HW_I2C0_SCL                 (9)
#define MICROPY_HW_I2C0_SDA                 (8)
```
------

In summary, all C3 and S3 boards already define default I2C0 pins *except ESP32_GENERIC_C3*.  So that board is the only one affected by this PR.

Since the majority of S3 boards define SCL/SDA to be 9/8 I suggest that's what this PR does for S3 in `machine_i2c.h`.

For C3... maybe it also makes sens to just use pins 9/8 as the default to match S3?  At least that makes the code a little simpler.

dpgeorge commented on ports/esp32/machine_i2c.c:
```diff
@@ -35,9 +35,14 @@
 #ifndef MICROPY_HW_I2C0_SCL
+#if CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32S3
+#define MICROPY_HW_I2C0_SCL (GPIO_NUM_7)
+#define MICROPY_HW_I2C0_SDA (GPIO_NUM_6)
```
As per my other comment, I suggest using 9 and 8 here.

ricksorensen replied:
With 18/19 as I2C pins:

-   Certainly the XIAO C3 does not behave with 18/19 and crashes REPL
-   The UM TINYS3, XIAO S3 keeps running even with the 19 conflict
-   I don't have C6 or S2 variants to test.

Note that the GENERIC C3, C6 boards do not have I2C pin defaults. And the only conflict with 8/9 is boot strapping use.
The GENERIC_S2 also does not define defaults but the GENERIC_S3 matches UM models.  The UM_C6 defaults are 6/7
 
I will change   the defaults for C3, C6, S2, and S3 to match SCL=9, SDA=8.  I do not believe the ESP32 can use these because of SPI for memory mapping.

FYI: Here are the GPIO pin note I used:

```
Notes on espressif pin usage.  Strapping pins are available for
general use after startup.

esp32: 0,2,5,12,15 strap
       6-11,16,17 SPI ext mem
       13,14 JTAG

s2: 0,45,46 strap
   39-42 JTAG
   26-32 SPI ext mem

s3: 0,3,45,46 strap
   19,20 USB-JTAG
   26-37 SPI ext mem

c3: 2,8,9 strap
    12-17 SPI ext mem
    18,19 USB/JTAG

c6: 4,5,8,9,15 strap
    12,13 US/JTAG
    24-30 SPI ext mem
```

dpgeorge replied:
Oh, I didn't think about the boot strapping pins.  It might be best to avoid those pins for I2C defaults then?

dpgeorge replied:
We should pick sensible defaults (which might be hard or not well defined...), and they will be used by all the ESP32_GENERIC* boards.

For non-generic boards, it will be up to the board definition whether it uses the defaults or defines custom defaults.

That said, I don't really know what "sensible defaults" are.  It's just that ESP32_GENERIC_S3 already uses 9/8 and we probably shouldn't change that.  We can have the defaults different for different variants if necessary.

ricksorensen commented:
The strapping impact is only at power on.  After boot these revert to
general GPIO pins.   When I do my simple test I use as SSD1306 terminal
which has power from the MCU, not independent.

The LOLIN C3 MINI uses 8/10 for example.

A board designer/user should take that into account and not connect a
device to hold the pin low (I think active low...).

For the non-ESP32 variants, maybe 6/7 or 10/11 look safe for all nonESP32
boards.

On Sun, Apr 13, 2025 at 9:16 PM Damien George ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In ports/esp32/machine_i2c.c
> <https://github.com/micropython/micropython/pull/17123#discussion_r2041306953>
> :
>
> > @@ -35,9 +35,14 @@
>  #if MICROPY_PY_MACHINE_I2C || MICROPY_PY_MACHINE_SOFTI2C
>
>  #ifndef MICROPY_HW_I2C0_SCL
> +#if CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32S3
> +#define MICROPY_HW_I2C0_SCL (GPIO_NUM_7)
> +#define MICROPY_HW_I2C0_SDA (GPIO_NUM_6)
>
> Oh, I didn't think about the boot strapping pins. It might be best to
> avoid those pins for I2C defaults then?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/micropython/micropython/pull/17123#discussion_r2041306953>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA5QPQBNUV4GGP5BXGABH3L2ZMK6RAVCNFSM6AAAAAB3BNCHWSVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDONRSHE3TGMJSG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Rick Sorensen
***@***.***
651-470-5511

ricksorensen commented:
Sorry- I have mucked this commit up terribly.  I will likely close and submit a new one tonight.

projectgus commented:
> Sorry- I have mucked this commit up terribly. I will likely close and submit a new one tonight.

You should still be able to rewrite this PR - even if you make a brand new local branch tonight and then "force push" it over the top of this existing PR branch. See also https://github.com/jimmo/git-and-micropython#README

ricksorensen committed:
esp32/machine_i2c: Fix default I2C pins for C3, S3.

The default I2C init does not require setting SCL or SDA but the default
I2C0 pins for C3, S3 conflict with the espressif GPIO usage.

For the C3, pins 18/19 are for USB/JTAG.  If used for I2C() they will cause
the REPL to hang on initialization of the I2C.

For the S3 pin 19 is allocated for USB/JTAG also but the defaults do not
seem to affect the REPL.

See related #16956.

Fixes issue #17103.

Signed-off-by: Rick Sorensen <rick.sorensen@gmail.com>
Signed-off-by: Angus Gratton <angus@redyak.com.au>

robert-hh commented on ports/esp32/machine_i2c.c:
```diff
@@ -35,8 +35,13 @@
+#if CONFIG_IDF_TARGET_ESP32
 #define MICROPY_HW_I2C0_SCL (GPIO_NUM_18)
 #define MICROPY_HW_I2C0_SDA (GPIO_NUM_19)
+#else
```
Does that apply to all targets except the standard ESP32, like S2, C6, ... as well? If it should apply to C3 and S3, then better make that change specific for these boards.

ricksorensen replied:
I will reverify today.  Only the C3/S3 have impacts with USB from my understanding of the espressif documentation.  

I only have C3 and S3 boards. However , it looks like _future_ devices from espressif idf 5.4.1 api documentation (GPIO Peripheral)
C5: GPIO 18/19 are SPI (flash I believe)
H2: GPIO 18/19 are SPI (flash I believe), GPIO 8/9 strapping
P4: No special usage for any of these pins

Decison for default pin seems arbitrary if the default initialization is desired for commonality.

OpEd: I don't like the default construct, as most designers will define their pin usage.  Conflicts with chip usage (e.g strapping, USB, .. ) are caveat emptor, but if default is used is should not impact other peripherals.

ricksorensen replied:
Should I update this PR? Just need architectural direction.

dpgeorge replied:
> I don't like the default construct, as most designers will define their pin usage.

Yes, I agree, having default pins for generic boards doesn't really work well because the default value is rather arbitrary.

But we are here now with default pins, so let's just try to make this work well enough.

I just noticed that the default pins for IC21 are already 9/8!  So maybe I2C0 should be different, maybe go back to 7/6 like it was originally in this PR?

I also think the defaults should be changed only for S3 and C3.  So probably the original PR was good, namely:
```c
#if CONFIG_IDF_TARGET_ESP32C3 || CONFIG_IDF_TARGET_ESP32S3
#define MICROPY_HW_I2C0_SCL (GPIO_NUM_7)
#define MICROPY_HW_I2C0_SDA (GPIO_NUM_6)
#else
...
#endif
```

ricksorensen commented:
Verified compilation against four ESP32 GENERIC Boards(ESP32, C3, S3_SPIRAM, S2) .  Tested on XIAO C3 and S3 boards.

Discovered S2 needed updated I2C(1) defaults during testing- no GPIO 25/26 available.

dpgeorge commented:
> Discovered S2 needed updated I2C(1) defaults during testing- no GPIO 25/26 available.

I think the I2C(1) defaults are fine as they were originally: only ESP32 uses 25/26, everything else uses 9/8.

projectgus commented:
@ricksorensen Let us know if you'd like a hand unpicking the changes in the branch, looks like something went a bit off.

ricksorensen commented:
Help would be greatly appreciated.  I believe the single file change
(machine_i2c.c) is good- but I must have updated my repository incorrectly,

projectgus commented:
@ricksorensen No worries. I've rebased and squashed all the commits together. The contents of esp32/machine_i2c.c is the same before and after the rebase (apart from one extra line of whitespace that I trimmed back out).
